### PR TITLE
Extracted ESP32 Url encoding method to the common QRCodeUtil module.

### PR DESCRIPTION
**Pull request by kkasperczyk-no**
#### Problem
ESP32 platform has implemented module enabling generating url link
to the webpage displaying QR code, what could be useful for other
platforms not having displays.

 #### Summary of Changes
* Moved url encoding method from ESP32 to the common QRCodeUtil module.
* Refactored methods a bit, as they were bug prone (e.g. unreserved
array was not /0 terminated and then strchr was using it).

### CODE REVIEW

File: examples/common/chip-app-server/include/QRCodeUtil.h [hunk 1]
```diff
@@ -21,3 +21,4 @@
 
 void PrintQRCode(chip::RendezvousInformationFlags rendezvousFlags);
 CHIP_ERROR GetQRCode(uint32_t & setupPinCode, std::string & QRCode, chip::RendezvousInformationFlags rendezvousFlags);
+CHIP_ERROR EncodeQRCodeToUrl(const char * QRCode, size_t len, char * url, size_t maxSize);
```
Comment (bzbarsky-apple) [inline examples/common/chip-app-server/include/QRCodeUtil.h:24]: Please document the arguments.

File: examples/common/chip-app-server/QRCodeUtil.cpp [hunk 2]
```diff
@@ -33,6 +37,15 @@ void PrintQRCode(chip::RendezvousInformationFlags rendezvousFlags)
     {
         ChipLogProgress(AppServer, "SetupPINCode: [%" PRIu32 "]", setupPinCode);
         ChipLogProgress(AppServer, "SetupQRCode:  [%s]", QRCode.c_str());
+
+        chip::Platform::ScopedMemoryBuffer<char> qrCodeBuffer;
+        const size_t qrCodeBufferMaxSize = 3 * QRCode.size() + 1;
+        qrCodeBuffer.Alloc(qrCodeBufferMaxSize);
+        if (EncodeQRCodeToUrl(QRCode.c_str(), QRCode.size(), &qrCodeBuffer[0], qrCodeBufferMaxSize) == CHIP_NO_ERROR)
+        {
+            ChipLogProgress(AppServer, "Copy paste the below URL in a browser to see the QR Code:\n\t%s?data=%s", qrCodeBaseUrl,
```
Comment (bzbarsky-apple) [inline examples/common/chip-app-server/QRCodeUtil.cpp:46]:
```suggestion
            ChipLogProgress(AppServer, "Copy/paste the below URL in a browser to see the QR Code:\n\t%s?data=%s", qrCodeBaseUrl,
```

File: examples/common/chip-app-server/QRCodeUtil.cpp [hunk 3]
```diff
@@ -75,3 +88,44 @@ CHIP_ERROR GetQRCode(uint32_t & setupPinCode, std::string & QRCode, chip::Rendez
...
+    return isalpha(c) || isdigit(c) || (strchr(specialCharsUnreservedInRfc3986, c) != nullptr);
+}
+
+CHIP_ERROR EncodeQRCodeToUrl(const char * QRCode, size_t len, char * url, size_t maxSize)
+{
+    const char upperHexDigits[] = "0123456789ABCDEF";
+    CHIP_ERROR err              = CHIP_NO_ERROR;
+    size_t i = 0, j = 0;
+
+    VerifyOrExit((QRCode && url) != NULL, err = CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (bzbarsky-apple) [inline examples/common/chip-app-server/QRCodeUtil.cpp:103]: Comparing a boolean to NULL is pretty weird. This should probably be `(QRCode != nullptr) && (url != nullptr)`

File: examples/common/chip-app-server/QRCodeUtil.cpp [hunk 4]
```diff
@@ -75,3 +88,44 @@ CHIP_ERROR GetQRCode(uint32_t & setupPinCode, std::string & QRCode, chip::Rendez
...
+    VerifyOrExit(len <= maxSize, err = CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (bzbarsky-apple) [inline examples/common/chip-app-server/QRCodeUtil.cpp:104]: This doesn't seem like a useful check.  `len < maxSize` does not guarantee that `maxSize` is big enough, right?  This just gives the impression of us having checked input length for sanity when we haven't really done that....

File: examples/common/chip-app-server/QRCodeUtil.cpp [hunk 5]
```diff
@@ -75,3 +88,44 @@ CHIP_ERROR GetQRCode(uint32_t & setupPinCode, std::string & QRCode, chip::Rendez
...
+
+            url[j++] = c;
+        }
+        else
+        {
+
+            VerifyOrExit((j + 3) < maxSize, err = CHIP_ERROR_BUFFER_TOO_SMALL);
+
+            url[j++] = '%';
+            url[j++] = upperHexDigits[c >> 4];
```
Comment (bzbarsky-apple) [inline examples/common/chip-app-server/QRCodeUtil.cpp:122]: I would really prefer it if this were `(c & 0xf0) >> 4` just to be safe.  Nothing says `char` can't represent values larger then 255, and then you'd potentially index out of the array here.

Yes, I know you just moved the code.

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by kkasperczyk-no** [comment]
@bzbarsky-apple I introduced changes related with all your remarks. Thanks for review and very relevant suggestions.

**Review by andy31415 — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

**Pull request merged**